pkg/migrate/cstor/snapshot.go

```diff
@@ -0,0 +1,186 @@
+	snap "github.com/openebs/maya/pkg/kubernetes/snapshot/v1alpha1"
+	snapData "github.com/openebs/maya/pkg/kubernetes/snapshotdata/v1alpha1"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+)
+
+// SnapshotMigrator ...
+type SnapshotMigrator struct {
+	pvName     string
+	snapClient *snapclientset.Clientset
+}
+
+var (
+	snapClass = "csi-cstor-snapshotclass"
+)
+
+func (s *SnapshotMigrator) migrate(pvName string) error {
+	s.pvName = pvName
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		return errors.Wrap(err, "error building kubeconfig")
+	}
+	// snapclientset.NewForConfig creates a new Clientset for VolumesnapshotV1beta1Client
+	s.snapClient, err = snapclientset.NewForConfig(cfg)
+	if err != nil {
+		klog.Fatalf("Failed to create snapshot client: %v", err)
+	}
+	return s.migrateSnapshots()
+}
+
+func (s *SnapshotMigrator) migrateSnapshots() error {
+	_, err := s.snapClient.SnapshotV1beta1().VolumeSnapshotClasses().
+		Get(snapClass, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to get snapshotclass %s", snapClass)
+	}
+	snapshotList, err := snap.NewKubeClient().
+		WithNamespace("").
+		List(metav1.ListOptions{
+			LabelSelector: "SnapshotMetadata-PVName=" + s.pvName,
+		})
+	if err != nil {
+		return err
+	}
+	for _, snapshot := range snapshotList.Items {
+		snapshot := snapshot // pin it
+		err = s.migrateSnapshot(&snapshot)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *SnapshotMigrator) migrateSnapshot(oldSnap *snapv1.VolumeSnapshot) error {
+	snapshotData, err := snapData.NewKubeClient().
+		Get(oldSnap.Spec.SnapshotDataName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to get volumesnapshotdata %s for %s", oldSnap.Spec.SnapshotDataName, oldSnap.Name)
+	}
+	klog.Infof("Creating equivalent volumesnapshotcontent for volumesnapshotdata %s", snapshotData.Name)
+	snapContent, err := s.createSnapContent(snapshotData, oldSnap)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create equivalent volumesnapshotcontent for volumesnapshotdata %s", snapshotData.Name)
+	}
+	klog.Infof("Creating equivalent new csi volumesnapshot for old volumesnapshot %s", oldSnap.Name)
+	newSnap, err := s.createNewSnapShot(snapContent, oldSnap)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create equivalent new csi volumesnapshot for old volumesnapshot %s", oldSnap.Name)
+	}
+	klog.Infof("Validating new csi volumesnapshot %s is bound to volumesnapshotcontent %s", newSnap.Name, snapContent.Name)
+	err = s.validateMigration(snapContent, newSnap)
+	if err != nil {
+		return errors.Wrapf(err, "failed to validate new volumesnapshot %s", newSnap.Name)
+	}
+	klog.Infof("Cleaing up old volumesnapshot %s", oldSnap.Name)
+	err = snap.NewKubeClient().WithNamespace(oldSnap.Namespace).Delete(oldSnap.Name, &metav1.DeleteOptions{})
```

Comment:
Add a note saying, users should perform migration only from latest version of OpenEBS.